contrib/log/slog/slog.go

```diff
@@ -31,21 +34,83 @@ func NewJSONHandler(w io.Writer, opts *slog.HandlerOptions) slog.Handler {
 
 // WrapHandler enhances the given logger handler attaching tracing information to logs.
 func WrapHandler(h slog.Handler) slog.Handler {
-	return &handler{h}
+	return &handler{wrapped: h}
 }
 
 type handler struct {
-	slog.Handler
+	wrapped    slog.Handler
+	groups     []string
+	groupAttrs map[string][]slog.Attr
+}
+
+// Enabled calls the wrapped handler Enabled method.
+func (h *handler) Enabled(ctx context.Context, level slog.Level) bool {
+	return h.wrapped.Enabled(ctx, level)
 }
 
 // Handle handles the given Record, attaching tracing information if found.
 func (h *handler) Handle(ctx context.Context, rec slog.Record) error {
+	reqHandler := h.wrapped
+
+	// We need to ensure the trace id and span id keys are set at the root level:
+	// https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/
+	// In case the user has created group loggers, we ignore those and
+	// set them at the root level.
 	span, ok := tracer.SpanFromContext(ctx)
 	if ok {
-		rec.Add(
-			slog.Uint64(ext.LogKeyTraceID, span.Context().TraceID()),
-			slog.Uint64(ext.LogKeySpanID, span.Context().SpanID()),
-		)
+		traceID := strconv.FormatUint(span.Context().TraceID(), 10)
+		spanID := strconv.FormatUint(span.Context().SpanID(), 10)
+
+		attrs := []slog.Attr{
+			slog.String(ext.LogKeyTraceID, traceID),
+			slog.String(ext.LogKeySpanID, spanID),
+		}
+		reqHandler = reqHandler.WithAttrs(attrs)
+	}
+	for _, group := range h.groups {
+		reqHandler = reqHandler.WithGroup(group)
+		if attrs, ok := h.groupAttrs[group]; ok {
+			reqHandler = reqHandler.WithAttrs(attrs)
+		}
+	}
+	return reqHandler.Handle(ctx, rec)
+}
+
+// WithAttrs saves the provided attributes associated to the current Group.
+// If Group was not called for the logger, we just call WithAttrs for the wrapped handler.
+func (h *handler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	if len(h.groups) == 0 {
+		return &handler{
+			wrapped:    h.wrapped.WithAttrs(attrs),
+			groupAttrs: h.groupAttrs,
+			groups:     h.groups,
+		}
+	}
+	curGroup := h.groups[len(h.groups)-1]
+
+	groupAttrs := groupAttrsCopy(h.groupAttrs)
+	groupAttrs[curGroup] = append(groupAttrs[curGroup], attrs...)
```

Comment:
It is possible that the same group name appears multiple times.
I think it might be better to keep the attributes by using the depth of groups.
like groupAttrs [][]slog.Attr

Reply:
@shota3506 you are right, nice catch! 👍